compiler/src/template-api/Table.js

```diff
@@ -0,0 +1,318 @@
+const Transform = require("../Transform").Transform;
+const Layer = require("../Layer").Layer;
+/*
+ * Constructor of a table
+ * @param args
+ * @constructor
+ * by xinli on 07/15/19
+ */
+
+function Table(args) {
+    if (args == null) args = {};
+
+    // check required args
+    var requiredArgs = ["name", "db", "fields"];
+    var requiredArgsTypes = ["string", "string", "object"];
+    for (var i = 0; i < requiredArgs.length; i++) {
+        if (!(requiredArgs[i] in args))
+            throw new Error(
+                "Constructing Table: " + requiredArgs[i] + " missing."
+            );
+        if (typeof args[requiredArgs[i]] !== requiredArgsTypes[i])
+            throw new Error(
+                "Constructing Table: " +
+                    requiredArgs[i] +
+                    " must be " +
+                    requiredArgsTypes[i] +
+                    "."
+            );
+        if (requiredArgsTypes[i] == "string")
+            if (args[requiredArgs[i]].length == 0)
+                throw new Error(
+                    "Constructing Table: " +
+                        requiredArgs[i] +
+                        " cannot be an empty string."
+                );
+    }
+
+    var table_name = "table_" + args.name;
+
+    var transform_func = getTableTransformFunc(table_name);
+
+    // schema and query first for buiding layer
+    var schema = args.fields.concat(["rn", "y"]);
+
+    var query = args.query || genQuery();
+
+    var transform = new Transform(query, args.db, transform_func, schema, true);
+
+    Layer.call(this, transform, false);
+
+    // TODO: check the type of arguments
+    this.cell_h = args.cell_h || 40;
+    this.x = args.x || 0;
+    this.y = args.y || 0;
+    this.db = args.db;
+
+    var sum_width = 0;
+    var centroid_x = 0;
+    if (typeof args.width === "number") {
+        sum_width = args.width;
+        centroid_x = this.x + args.width / 2;
```

Comment:
minor: this statement to calculate `centroid_x` is repeated in each if/else branch, can be put after this if/else

Reply:
Sure